zebra-consensus/src/script.rs

```diff
@@ -12,11 +12,181 @@
 //! This is an internal module. Use `verify::BlockVerifier` for blocks and their
 //! transactions, or `mempool::MempoolTransactionVerifier` for mempool transactions.
 
+use std::{pin::Pin, sync::Arc};
+
+use std::future::Future;
+use zebra_chain::{parameters::ConsensusBranchId, transaction::Transaction, transparent};
+
+use crate::BoxError;
+
 /// Internal script verification service.
 ///
 /// After verification, the script future completes. State changes are handled by
 /// `BlockVerifier` or `MempoolTransactionVerifier`.
 ///
 /// `ScriptVerifier` is not yet implemented.
-#[derive(Default)]
-pub(crate) struct ScriptVerifier {}
+pub struct Verifier<ZS> {
+    state: ZS,
+    branch: ConsensusBranchId,
+}
+
+impl<ZS> Verifier<ZS> {
+    pub fn new(state: ZS, branch: ConsensusBranchId) -> Self {
+        Self { state, branch }
+    }
+}
+
+#[derive(Debug)]
+struct Request {
+    transaction: Arc<Transaction>,
+    input_index: usize,
+}
+
+impl<ZS> tower::Service<Request> for Verifier<ZS>
+where
+    ZS: tower::Service<zebra_state::Request, Response = zebra_state::Response, Error = BoxError>,
+    ZS::Future: Send + 'static,
+{
+    type Response = ();
+    type Error = BoxError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.state.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request) -> Self::Future {
+        use futures_util::FutureExt;
+
+        let input = &req.transaction.inputs()[req.input_index];
+
+        match input {
+            transparent::Input::PrevOut { outpoint, .. } => {
+                let output = self.state.call(zebra_state::Request::AwaitUtxo(*outpoint));
+                let transaction = req.transaction;
+                let branch_id = self.branch;
+                let input_index = req.input_index;
+
+                async move {
+                    let previous_output = match output.await? {
+                        zebra_state::Response::Utxo(output) => output,
+                        _ => unreachable!("AwaitUtxo always responds with Utxo"),
+                    };
+
+                    zebra_script::is_valid(
+                        transaction,
+                        branch_id,
+                        (input_index as u32, previous_output),
+                    )?;
+
+                    Ok(())
+                }
+                .boxed()
+            }
+            transparent::Input::Coinbase { .. } => unimplemented!(
+                "how should we handle verifying coinbase transactions in the script::Verifier?"
```

Comment:
> The block height MUST be encoded as the first item in the coinbase transaction’s scriptSig, as specified in [BIP-34]. The format of the height is “serialized CScript” – the first byte is the number of bytes in the number, and the following bytes are the signed little-endian representation of the number.

https://zips.z.cash/protocol/canopy.pdf#blockheader

Reply:
I think the async script verifier should check the script part of this rule, and then the subsidy module should check the address:
> Consensus rule:[Pre-Canopy]  A coinbase transaction at height ∈ {1..FoundersRewardLastBlockHeight} MUST include at least one output that pays exactly FoundersReward(height) zatoshi with a standard P2SH script of the form OP_HASH160 FounderRedeemScriptHash(height) OP_EQUAL as its scriptPubKey.

https://zips.z.cash/protocol/canopy.pdf#foundersreward

Reply:
Similarly with transparent founders rewards:

> Consensus rule: [Canopy onward]  The coinbase transaction at block height `height` MUST contain at least one output per funding stream `fs` active at `height`, that pays `fs.Value(height)` zatoshi in the prescribed way to the stream’s recipient address represented by `fs.AddressListfs.AddressIndex(height)`.
> The “prescribed way” to pay a transparent P2SH address is to use a standard P2SH script of the form OP_HASH160 `fs.RedeemScriptHash(height)` OP_EQUAL as the scriptPubKey. Here `fs.RedeemScriptHash(height)` is the standard redeem script hash for the recipient address given by `fs.AddressList[fs.AddressIndex(height)]` in Base58Check form. The standard redeem script hash is specified in [Bitcoin-Multisig] for P2SH multisig addresses, or [Bitcoin-P2SH] for other P2SH addresses.

Reply:
To be clear, I think we can merge this PR now, and open a ticket for coinbase script validation.

Reply:
I don't think that this is the appropriate place to check this rule, because this rule isn't actually a rule scoped to a particular script input, but a rule scoped to a block: each block has exactly one transaction with coinbase inputs, of such and such form, etc.  

So, I think that a better strategy that keeps the script verifier scoped to script verification would be to handle the special coinbase transaction specially, and reject coinbase inputs in the script verifier (which would then be scoped to ordinary transactions).